presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -59,19 +60,27 @@ default boolean isEnabled(Session session)
     {
         public static Result empty()
         {
-            return new Result(Optional.empty());
+            return new Result(Optional.empty(), Optional.empty());
         }
 
         public static Result ofPlanNode(PlanNode transformedPlan)
         {
-            return new Result(Optional.of(transformedPlan));
+            return new Result(Optional.of(transformedPlan), Optional.empty());
+        }
+
+        public static Result setTrait(Trait trait)
+        {
+            return new Result(Optional.empty(), Optional.of(trait));
         }
 
         private final Optional<PlanNode> transformedPlan;
+        private final Optional<Trait> trait;
 
-        private Result(Optional<PlanNode> transformedPlan)
+        private Result(Optional<PlanNode> transformedPlan, Optional<Trait> trait)
         {
             this.transformedPlan = requireNonNull(transformedPlan, "transformedPlan is null");
+            this.trait = requireNonNull(trait, "trait is null");
+            checkArgument(!(transformedPlan.isPresent() && trait.isPresent()), "Either transformed plan or trait can be set");
```

Comment:
So far I wanted to be on the safe side, so I wanted to prevent anything that is not expected to happen today.

For me, your use case sounds ok. If a rule is capable to tell traits its is better. I think we can easily remove that restriction if needed. Would you like me to change it now?

Reply:
@kokosing Grzegorz, thanks for a quick response. Let's not make any changes just yet. I'd like to read through all the commits first and, if you don't mind, I'd like to ask questions as I go.